### ISSUE TITLE
Wrong parameters for updateGradInput

### ISSUE THREAD

**Issue by shawnjhenry** [soumith/dcgan.torch#21]
Hi,

I'm attempting to use the GAN framework in a different setting than images and I was using this code as a reference.  I noticed something odd:  When we call optim.adam on fDx, the parameters in netD get updated, but we use the output with respect to the original parameters when we call it on fGx.  Shouldn't we call forward on netD inside fGx rather than recycle the previous output so that both output and gradInput are computed with respect to the same parameters?

Thanks,
Shawn

**Comment by Guim3** [comment]
It doesn't make much of a difference. You would only simply give the discriminator one training iteration extra training compared to the generator. And in fact, you are interested on not overpowering the discriminator, as this might yield an unstable training. 
Think it like this way: if you were to compute the output with the updated netD, why wouldn't you update the parameters again with this new output? You could add as many extra iterations you wanted to the discriminator for one iteration from the generator, but this doesn't might be desired for a robust model.

**Comment by shawnjhenry** [comment]
Unless I'm misunderstanding something, I don't think that's what happens in the code.  What happens in the code is that we call forward on netD in fDx, then update netD's parameters, then call backward on netD in fGx.  That means that the gradInput is being computed using the updated parameters, while the loss was computed using the parameters before they were updated.  That means the gradient doesn't actually make any sense at all.

**Comment by Guim3** [comment]
Hi again,

I misunderstood your question, sorry! 
When you say "then call backward on netD in fGx", you are refering [to this line of code](https://github.com/soumith/dcgan.torch/blob/master/main.lua#L200), right? 

`local df_dg = netD:updateGradInput(input, df_do)`

Firstly, notice that a call to [updateGradInput](https://github.com/torch/nn/blob/master/doc/module.md#updategradinputinput-gradoutput) _does not_ compute a backpropagation step. In order to do so, you would also need to call accGradParameters, which is not called on D in fGx. You can try this by calling multiple times updateGradInput, you will see that the gradient returned is always the same.

Secondly, why is necessary to call updateGradInput on D if we are not updating D at all? Because we need the gradients with respect to the fake images generated by G (`input` in fGx). This is what updateGradInput returns, `df_dg`.

So, once we get `df_dg`, we are now able to do a backward step on G with the correct gradients:

`netG:backward(noise, df_dg)`

PS: [Notice that `df_do`](https://github.com/soumith/dcgan.torch/blob/master/main.lua#L199) does not contain the gradients of the generated images `input` but the gradients of D output, which is a one dimensional vector. 

Hope I have answered your question this time! :P

**Comment by shawnjhenry** [comment]
Sorry, I misspoke - I meant we call updateGradInput on netD in fGx, but that doesn't address my point at all.  The problem is that netD had different parameters when we call forward on it in fDx than when we call updateGradInput on it in fGx since we ran adam on fDx before calling fGx.  That means that we are taking the gradient of netD's input using different parameters than we used to compute the output of netD, so the loss and gradient being computed don't match, although they will be close assuming the parameter update is small, which is why the model still works even though the gradient computation is incorrect.  I've corrected it in my own code and it doesn't make a huge difference, but I would still recommend correcting it here or adding a note that it is not technically correct.

Maybe a pseudo code simplification of what's going on will make this more clear:

1.) fDx
z <- N(0, 1)
x_real <- load from dataset 
x_fake <- netG(z)
y_real <- netD(x_real)
...criterion stuff...
...gradInput stuff...
gradParamsD <- dnetD_dparamsD
y_fake <- netD(x_fake)
...criterion stuff...
...gradInput stuff...
gradParamsD <- dnetD_dparamsD 

2.) adam 

paramsD <- adam(gradParamsD) 

3.) fGx 

...criterion stuff using y_fake from 1.) ... 
gradInputD <- dnetD_dx_fake (the params of netD have changed in 2.) since we computed y_fake!) 
...netG stuff using gradInputD...

4.) adam 

paramsG <- adam(gradParamsG)

**Comment by Guim3** [comment]
This time I think I got your point.
I agree with you.
Then you would simply change [this line](https://github.com/soumith/dcgan.torch/blob/master/main.lua#L197):

`local output = netD.output -- netD:forward(input) was already executed in fDx, so save computation`

For this one:

`local output = netD:forward(input)`

It remains to be seen if the time saved by inaccurately updating the parameters pays off, though.

**Comment by shawnjhenry** [comment]
Yeah, that's sort of what I did.  It's actually possible to do it correctly without doing netD:forward(input) - what you do is create a closure that does the full computation for both netD and netG and returns as its result a pair of functions that return the D loss and gradParams and G loss and gradParams separately.  Then in training we call the single closure and then apply adam to functions it returns.  This easily generalizes for multiple G updates per D update if desired.  To give a better idea we create:

```
function trainBatch()
    ...do all computations...

    local function trainBatchD(p) return lossD, gradParamsD end
    local function trainBatchG(p) return lossG, gradParamsG end

    return trainBatchD, trainBatchG 
end

then in training we do:

local trainBatchD, trainBatchG = trainBatch()

_, lossD = optim.adam(trainBatchD, paramsD, optimStateD)
_, lossG = optim.adam(trainBatchG, paramsG, optimStateG)
```
